themes/default/content/blog/infrastructure-testing-concepts/index.md

```diff
@@ -0,0 +1,95 @@
+
+![Testing rocket](testing_rocket.png)
+
+For deploying modern applications with cloud engineering, we propose an alternate model called the testing rocket. Static testing is inherent in cloud engineering because of the software toolchains, such as IDEs, that perform static checking. Both unit and integration testing are equally importat to ensure that infrastructure is deployed and managed reliabily. End-to-end tests for performance and scalability should be performed less frequently because of complexity and costs.
+
+## Cloud engineering testing
+
+Cloud engineering applies software testing principles to infrastructure where we focus on functional testing. We use three types of functional tests with infrastructure: unit tests, integration tests, and property tests (sometimes called functional tests). We'll examine each type of test in detail.
+
+## Static tests
+
+A word about static tests and infrastructure. Infrastructure deployed with templating languages such as YAML or JSON are limited to linting, i.e., validating and formatting the code. Deploying infrastructure with programming languages lets you use built-in tools in IDEs to perform static tests that highlight errors in your code in addition to features such as code completion, enumerations, and syntax checking. Pulumi's preview also performs static checking before deploying a resource.
+
+### Unit tests
+
+Unit tests for applications test the smallest piece of code, which are often methods and functions of classes or modules. They are commonly written by the programmer and run quickly.
+
+With infrastructure, the smallest unit is often the cloud resource. Because cloud resources are not created, you can’t write a test to evaluate infrastructure behavior in a templating language. For example, you can’t make HTTP requests to endpoints because there’s no web server to serve them.
+
+Because cloud engineering is software engineering, Pulumi replaces all external dependencies with [mock objects](https://en.wikipedia.org/wiki/Mock_object) that replicate the behavior in a specified way. Unit tests are very fast because they run in memory without any out-of-process calls. They provide fast feedback loops during development, making them suitable for Test-Driven Development (TDD).
+
+As with applications, unit tests for infrastructure in Pulumi are authored with the same language used to declare infrastructure.  This means you can use familiar test and mock frameworks such as PyTest for Python or Mocha for Node.js.  Using a framework ensures configuration is correct before provisioning and that the resulting infrastructure has the specified properties. In addition, team standards and security guidelines are enforced.
```

Comment:
```suggestion
Unit tests for applications test the smallest piece of code, which are often methods and functions of classes or modules. They are commonly written by the programmer and run quickly.

With infrastructure, the smallest unit is often the cloud resource. Because cloud resources haven't been created by the time you run unit tests, you can’t write a test to evaluate infrastructure behavior. For example, you can’t make HTTP requests to endpoints because there’s no web server to serve them.

Pulumi's cloud engineering approach enables unit testing by replacing all external dependencies with [mock objects](https://en.wikipedia.org/wiki/Mock_object) that replicate the behavior in a specified way. This enables you to write fast unit tests that run in memory without any out-of-process calls. They provide rapid feedback loops during development, making them suitable for Test-Driven Development (TDD).

As with applications, unit tests for infrastructure in Pulumi are authored with the same language used to declare infrastructure. This means you can use familiar test and mock frameworks such as PyTest for Python or Mocha for Node.js. Using a framework ensures configuration is correct before provisioning and that the resulting infrastructure has the specified properties. In addition, team standards and security guidelines are enforced.
```